Neos.ContentGraph.DoctrineDbalAdapter/src/ContentRepositoryReadModelAdapter.php

```diff
@@ -0,0 +1,209 @@
+use Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceName;
+use Neos\ContentRepository\Core\SharedModel\Workspace\Workspaces;
+use Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceStatus;
+use Neos\EventStore\Model\Event\Version;
+
+/**
+ * @internal only used inside the
+ * @see ContentRepositoryReadModel
+ */
+final readonly class ContentRepositoryReadModelAdapter implements ContentRepositoryReadModelAdapterInterface
+{
+    public function __construct(
+        private Connection $dbal,
+        private NodeFactory $nodeFactory,
+        private ContentRepositoryId $contentRepositoryId,
+        private NodeTypeManager $nodeTypeManager,
+        private ContentGraphTableNames $tableNames
+    ) {
+    }
+
+    public function buildContentGraph(WorkspaceName $workspaceName, ContentStreamId $contentStreamId): ContentGraph
+    {
+        return new ContentGraph($this->dbal, $this->nodeFactory, $this->contentRepositoryId, $this->nodeTypeManager, $this->tableNames, $workspaceName, $contentStreamId);
+    }
+
+    public function findWorkspaceByName(WorkspaceName $workspaceName): ?Workspace
+    {
+        $workspaceByNameStatement = <<<SQL
+            SELECT
+                name, baseWorkspaceName, currentContentStreamId, status
+            FROM
+                {$this->tableNames->workspace()}
+            WHERE
+                name = :workspaceName
+            LIMIT 1
+        SQL;
+        try {
+            $row = $this->dbal->fetchAssociative($workspaceByNameStatement, [
+                'workspaceName' => $workspaceName->value,
+            ]);
+        } catch (Exception $e) {
+            throw new \RuntimeException(sprintf('Failed to load workspace from database: %s', $e->getMessage()), 1716486077, $e);
+        }
+        if ($row === false) {
+            return null;
+        }
+        return self::workspaceFromDatabaseRow($row);
+    }
+
+    public function findWorkspaces(): Workspaces
+    {
+        $workspacesStatement = <<<SQL
+            SELECT
+                name, baseWorkspaceName, currentContentStreamId, status
+            FROM
+                {$this->tableNames->workspace()}
+        SQL;
+        try {
+            $rows = $this->dbal->fetchAllAssociative($workspacesStatement);
+        } catch (Exception $e) {
+            throw new \RuntimeException(sprintf('Failed to load workspaces from database: %s', $e->getMessage()), 1716902981, $e);
+        }
+        return Workspaces::fromArray(array_map(self::workspaceFromDatabaseRow(...), $rows));
+    }
+
+    public function findContentStreamById(ContentStreamId $contentStreamId): ?ContentStream
+    {
+        $contentStreamByIdStatement = <<<SQL
+            SELECT
+                id, sourceContentStreamId, status, version
+            FROM
+                {$this->tableNames->contentStream()}
+            WHERE
+                id = :contentStreamId
+                AND removed = FALSE
```

Comment:
I rethought this again and again and i think the `removed` state is not just an implementation detail but required to be implemented by the adapter. That means we can and should add it to the `ContentStream` readmodel so it is clearly documented when this state should be encountered.
Further adding it to the readmodel allows us to remove the `findUnusedAndRemovedContentStreamIds` in the adapters and interfaces which would be impossible to document due to their complexity. It really feels good to have that now in the ContentStreamPrunner in the core.

see 041f23926f41fa0ad2dad6e62f16efb45aa2e836

Reply:
After discussing this again we found a way to get rid of the removed flag from the readmodel: by diffing the actual found content streams of the event store and the ones that are in the projection. And as a second step im currently fully overhauling the content stream pruner so that we dont have to track neither the removed nor the content stream state at all in the projection and always extract the information from the events: https://github.com/neos/neos-development-collection/pull/5297

So the introduction of `ContentStream::isRemoved` was just a temporary state and will be removed again :)